framework/src/media/audio/audio_manager.c

```diff
@@ -181,54 +233,82 @@ audio_manager_result_t find_output_audio_card(void)
 	return AUDIO_MANAGER_SUCCESS;
 }
 
-audio_manager_result_t set_audio_volume(uint16_t volume)
+static audio_manager_result_t set_audio_volume(audio_device_type_t type, audio_config_t *audio_config, uint8_t volume)
 {
 	int ret;
+	int scaled_volume;
 	struct pcm *pcm;
+	struct audio_caps_desc_s caps_desc;
+	FAR struct audio_caps_s *caps = &(caps_desc.caps);
 
-	if ((ret = get_active_audio_device_pcm(&pcm, OUTPUT)) != AUDIO_MANAGER_SUCCESS) {
+	if ((ret = get_active_audio_device_pcm(&pcm, type)) != AUDIO_MANAGER_SUCCESS) {
 		return ret;
 	}
 
-	struct audio_caps_desc_s caps_desc;
-	FAR struct audio_caps_s *caps = &(caps_desc.caps);
-	ret = ioctl(pcm_get_file_descriptor(pcm), AUDIOIOC_GETCAPS, (unsigned long)caps);
-	if (ret < 0) {
-		auddbg("Fail to ioctl AUDIOIOC_CONFIGURE, ret = %d\n", ret);
-		return AUDIO_MANAGER_DEVICE_FAIL;
+	if (volume > AUDIO_DEVICE_MAX_VOLUME) {
+		volume = AUDIO_DEVICE_MAX_VOLUME;
 	}
 
-	caps->ac_controls.hw[0] = volume;
+	caps->ac_len = sizeof(struct audio_caps_s);
 
 	caps->ac_type = AUDIO_TYPE_FEATURE;
-	caps->ac_format.hw = AUDIO_FU_VOLUME;
+
+	if (type == INPUT) {
+		caps->ac_format.hw = AUDIO_FU_INP_GAIN;
+	} else {
+		caps->ac_format.hw = AUDIO_FU_VOLUME;
+	}
+
+	scaled_volume = volume * audio_config->max_volume / AUDIO_DEVICE_MAX_VOLUME;
+	caps->ac_controls.hw[0] = scaled_volume;
 
 	ret = ioctl(pcm_get_file_descriptor(pcm), AUDIOIOC_CONFIGURE, (unsigned long)&caps_desc);
 	if (ret < 0) {
 		auddbg("Fail to set a volume, ret = %d\n", ret);
 		return AUDIO_MANAGER_DEVICE_FAIL;
 	}
 
+	audio_config->volume = volume;
+	audvdbg("Volume = %d\n", audio_config->volume);
+
 	return AUDIO_MANAGER_SUCCESS;
 }
 
-int get_audio_volume(void)
+audio_manager_result_t set_input_audio_volume(uint8_t volume)
 {
-	int ret;
-	int volume = -1;
-	struct pcm *pcm;
+	audio_config_t *audio_config = &(g_audio_in_cards[g_actual_audio_in_card_id].config);
 
-	if ((ret = get_active_audio_device_pcm(&pcm, OUTPUT)) != AUDIO_MANAGER_SUCCESS) {
-		return ret;
+	if (audio_config->volume == volume) {
+		return AUDIO_MANAGER_SUCCESS;
 	}
 
-	ret = ioctl(pcm_get_file_descriptor(pcm), AUDIOIOC_GETVOLUME, (unsigned long)&volume);
-	if (ret < 0) {
-		auddbg("Fail to ioctl AUDIOIOC_GETVOLUME, ret = %d\n", ret);
-		return AUDIO_MANAGER_DEVICE_FAIL;
+	return set_audio_volume(INPUT, audio_config, volume);
+}
+
+audio_manager_result_t set_output_audio_volume(uint8_t volume)
+{
+	audio_config_t *audio_config = &(g_audio_out_cards[g_actual_audio_out_card_id].config);
+
+	if (audio_config->volume == volume) {
+		return AUDIO_MANAGER_SUCCESS;
 	}
 
-	return volume;
+	return set_audio_volume(OUTPUT, audio_config, volume);
+}
+
+int get_input_audio_volume(void)
+{
+	return g_audio_in_cards[g_actual_audio_in_card_id].config.volume;
```

Comment:
No, it should get from driver.

Reply:
If another application adjust through the ioctl without audiomanger, this will be wrong